Archival bucketlist

SirTyson opened the pull request:
# Description

Resolves #4394

Depends on XDR changes in https://github.com/stellar/stellar-xdr/pull/200.

This change refactors the BucketList so that we can have multiple Bucket Lists simultaneously. This is necessary for State Archival, where validators will maintain a live BucketList, hot archive BucketList, and cold archive BucketList. This change does not yet support writing the new archival BucketLists to history, 

Each BucketList has small differences wrt to the entry types it stores, merge logic, and how lookups are done, but the overall BucketList level logic is the same. Because of this, it seemed easiest to template the Bucket related classes and specialize a few functions. The difference are as follows (I'll be updating the Archival CAP with this info soon)

Entry Types:
Live BucketList: `BucketEntry`
Hot Archive BucketList: `HotArchiveBucketEntry`.

This change was necessary due to how LedgerKeys are stored. In the Live BucketList, only `DEADENTRY` store plain `LedgerKey`. `DEADENTRY` is equivalent to null, so these LedgerKeys are dropped at the bottom level bucket. 

Hot Archives require two types that store LedgerKeys: `HA_LIVE` and `HA_DELETED`. `HA_LIVE` is equivalent to the "null" type in the Live BucketList, and is dropped in the bottom bucket. However, `HA_DELETED` needs to be persisted. 

Merging:
In the hot archive, `LedgerEntry` are never updated. They may be overwritten by another `HotArchiveBucketEntry` type, but the `LedgerEntry` contents itself never change. This means that the `INITENTRY`, `LIVEENTRY`, and `DEADENTRY` abstraction doesn't really make sense. This makes the merge logic for Hot Archive buckets very simple, the newer entry always overwrites the older entry.

This PR adds the concept of a "Tombstone" entry. A tombstone entry essentially represents null in the given BucketList and can be dropped once it reaches the bottom bucket. On the live BucketList, DEADENTRY is the tombstone type. On the Hot Archive BucketList, `HA_LIVE` is the tombstone.

BucketListDB lookup:
The live BucketList only ever returns `LedgerEntry` types when queried. Any DEADENTRY LedgerKey is simply omitted from the return, as this is the "null" tombstone type.

Hot archive lookups must return both LedgerKey and LedgerEntry types. HA_LIVE entries are of type LedgerKey and can be omitted from the return since they are the tombstone type. However, HA_ARCHIVED entries are not tombstone entries and must be returned when found. Do to this, Hot Archive lookups return `HotArchiveBucketEntry` instead of `LedgerEntry` when queried.

Future updates will add more functionality to Hot Archives, further distinguishing it from the live BucketList, and add a third BucketList type called the cold archive. This cold archive will add a third BucketEntry type and have different merge logic as well, so I think the templated Bucket classes, while verbose, are warranted.

This is currently a draft until the XDR changes are merged.
 

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

dmkozh commented on src/bucket/BucketList.cpp:
```diff
@@ -24,71 +24,91 @@
+template <> BucketListDepth BucketListBase<LiveBucket>::kNumLevels = 11;
+
+// TODO: I made this number up, do some analysis and pick a better value or
+// make this a configurable network config.
```
Just curious, can we actually change this value at runtime in-between ledgers?

SirTyson replied:
No. We will probably end up making this a config setting at which point I'll change it to a dynamic runtime value, but for now it's easier just to inherit the older BucketList's global const value for this refactor.

dmkozh commented on src/bucket/BucketList.cpp:
```diff
@@ -510,16 +565,98 @@ BucketList::getSize() const
+
+    for (uint32_t i = static_cast<uint32>(mLevels.size()) - 1; i != 0; --i)
+    {
+        /*
```
Don't forget to remove this (or uncomment if that's universally useful for debuggin)

dmkozh commented on src/bucket/BucketManagerImpl.cpp:
```diff
@@ -948,7 +1088,23 @@ BucketManagerImpl::snapshotLedger(LedgerHeader& currentHeader)
+        if (protocolVersionStartsFrom(currentHeader.ledgerVersion,
+                                      ProtocolVersion::V_22))
+        {
+            // TODO: Hash Archive Bucket
```
What's the status of this TODO? I.e. is the PR blocked on this?

dmkozh commented on src/bucket/BucketManagerImpl.cpp:
```diff
@@ -1148,26 +1304,30 @@ BucketManagerImpl::assumeState(HistoryArchiveState const& has,
     releaseAssertOrThrow(mApp.getConfig().MODE_ENABLES_BUCKETLIST);
 
-    for (uint32_t i = 0; i < BucketList::kNumLevels; ++i)
+    // TODO: Assume archival bucket state
```
What about this TODO?

SirTyson replied:
The outstanding TODOs are outside the scope of this PR. I've added some additional context on their dependencies. They require changing the history archive structure, which is a pretty big endeavor.

dmkozh replied:
Let's maybe open an issue or a few issues for these - in general, Core codebase avoids TODOs merged, and if we do have some, I would prefer to have them point at the issues.

SirTyson replied:
We do, the master issue has most of these covered in sub-issues: https://github.com/stellar/stellar-core/issues/4392. I agree it's a little messy but there's a lot of changes that are all kinda have dependencies and I'm trying to merge them in small pieces

dmkozh commented on src/bucket/BucketSnapshot.cpp:
```diff
@@ -125,26 +138,40 @@ BucketSnapshot::loadKeysWithLimits(std::set<LedgerKey, LedgerEntryIdCmp>& keys,
+                    else
                     {
-                        result.push_back(entryOp->liveEntry());
+                        result.push_back(*entryOp);
```
I'm not sure I understand this, why do we add entry unconditionally as it seems like it must have been archived at this point? Is that for `getLedgerEntry` endpoint?

SirTyson replied:
Archival is not enforced at the database level, but at the txn level. Core needs to query about Archival State for a variety of reasons. For example, to check if a new entry creation is valid, core needs to check that an archived version of that key does not exist in the Live bucketlist or the archival bucketlists, so we can't just return null whenever an entry is archived at the db level.

dmkozh commented on src/bucket/BucketSnapshot.h:
```diff
@@ -37,32 +49,44 @@ class BucketSnapshot : public NonMovable
+                  std::vector<BulkLoadReturnT>& result,
+                  LedgerKeyMeter* lkMeter) const;
+
+    // friend struct BucketLevelSnapshot<BucketT>;
```
Please remove

dmkozh commented on src/bucket/LedgerCmp.h:
```diff
@@ -126,10 +129,70 @@ struct LedgerEntryIdCmp
+            if (aty != HOT_ARCHIVE_DELETED && aty != HOT_ARCHIVE_LIVE)
+            {
+                throw std::runtime_error(
+                    "Malformed bucket: expected DELETED/RESTORED key.");
```
We don't have 'RESTORED' key category, do we?

dmkozh commented on src/util/ProtocolVersion.h:
```diff
@@ -33,7 +33,8 @@ enum class ProtocolVersion : uint32_t
     V_20,
-    V_21
+    V_21,
+    V_22
```
Don't forget to update this to include v23 and use it where appropriate.
I also would suggest using a named constant instead of the direct version checks in order to make changes easier.

marta-lokhova commented:
Note: changes in this PR impact BucketListDB quite a bit, I think we need to resolve https://github.com/stellar/stellar-core/issues/4433 to make sure we have good coverage. Wdyt?

dmkozh commented on src/history/HistoryArchive.h:
```diff
@@ -27,13 +27,15 @@ namespace stellar
     std::string curr;
-    FutureBucket next;
+
+    // TODO: Add archival buckets to history
```
TODO

dmkozh reviewed (commented):
LGTM on the surface level. The code looks fine for now. It would be nicer if this PR was smaller, but I'm not sure if there is a way to stage this change. Rebasing should at least help getting rid of meta diffs.

SirTyson commented on src/bucket/Bucket.h:
```diff
@@ -96,6 +95,56 @@ class Bucket : public std::enable_shared_from_this<Bucket>,
+};
+
+template <class BucketT> class SearchableBucketListSnapshot;
+class LiveBucket : public Bucket,
```
Added some comments and rebased.

graydon reviewed (changes requested):
Sorry for slow review. Mostly all makes sense and looks .. well, a little gnarly, but good! Especially in light of recent chat.

Requested changes fall into 3 groups:
  1. A handful of "definite" changes (like I think one or two little typo-sized bugs)
  2. A handful of "maybe you could refactor/dedupe this new duplication" cases
  3. An even-more-general / open-ended "is this really the best way to do the type parameterization" questions.
 
I'd like to at least see cases 1 fixed and 2 attempted (or explained why not to, if you have already tried and the result is worse). Case 3 is more a latent request for discussion / exploration of the alternative parameterization patterns, like maybe it all relates to the impl pattern or something? I'm curious but won't block on the issue either way. I can also try some sketching alternative parameterization patterns on my own.

I guess I would like, if nothing else, for all the function bodies that contain a branch on "if-constexpr type-A-or-B" to also contain the "static-assert-type-is-either-A-or-B" call (and for it to be factored out into a named helper). It seems to be true in _many_ cases, but I'd like it to at least be true in all cases where the if-constexpr thing is employed. Just like so we can get a static list of places to fix when A-or-B becomes A-or-B-or-C in the future.

SirTyson commented on src/bucket/Bucket.cpp:
```diff
@@ -794,24 +935,43 @@ Bucket::merge(BucketManager& bucketManager, uint32_t maxProtocolVersion,
+    // well
+    if (ni.getMetadata().ext.v() == 1)
+    {
+        releaseAssertOrThrow(protocolVersionStartsFrom(
```
So the relevant function here is `calculateMergeProtocolVersion`. The current protocol version always gets fed in as `maxProtocolVersion` in `Bucket::merge` (the relevant entry points are the `FutureBucket` constructions in `BucketLevel<BucketT>::prepare`). `calculateMergeProtocolVersion` just takes the max protocol of the input buckets, then makes sure it's not greater than `maxProtocolVersion`. The funkiness comes in when shadows are involved, which will never happen in the archival buckets.

SirTyson replied:
Hmm so I can't actually use a `BucketT` interface for futures either, but I can still have seperate storage types in `BucketManager` to remove the case. I think I can get rid of all the `constexpr` branching by changing the getBucket and getFutures interface to static getter. It's not a very elegant interface, but avoids removing the `impl` construct and removes most constexpr branching, so it's probably the least evil approach.

SirTyson commented on src/bucket/BucketSnapshot.cpp:
```diff
@@ -8,66 +8,77 @@
     if (isEmpty())
     {
-        return {std::nullopt, false};
+        return {nullptr, false};
```
This is to avoid an extra copy. Previously, `SearchableBucketListSnapshot::load` was the high-level entry point for point loads and returned `std::shared_ptr<LedgerEntry>`. This lower level function, `getEntryAtOffset`, returned a `BucketEntry`, and we needed to construct a `LedgerEntry` anyway, so returning optional was fine.

However, now the Hot Archive BucketList returns a `HotArchiveBucketEntry` type. If we kept this lower level interface as an optional, we'd have to make a copy of the optional contents to construct the shared_ptr of HotArchiveBucketEntry returned by the higher level interface.

Alternatively, we could switch the higher level interfaces of `SearchableLiveBucketListSnapshot::load` and `SearchableHotArchiveBucketListSnapshot::load` to return optionals instead of shared_ptr, but that breaks some of the ltx cache that relies on shared_ptr.

graydon replied:
Hmm, I'm not 100% clear on this but it's not a big deal.

SirTyson commented on src/bucket/BucketSnapshot.cpp:
```diff
@@ -101,32 +113,65 @@ BucketSnapshot::loadKeysWithLimits(std::set<LedgerKey, LedgerEntryIdCmp>& keys,
     while (currKeyIt != keys.end() && indexIter != index.end())
     {
+        // lkMeter only supported for LiveBucketList
+        if (std::is_same_v<BucketT, LiveBucket> && lkMeter)
```
Ah this was the result of a bad rebase, I'll remove it.

SirTyson commented on src/bucket/test/BucketListTests.cpp:
```diff
@@ -337,58 +386,118 @@ TEST_CASE_VERSIONS("bucket list shadowing pre/post proto 12",
                 bm, getAppLedgerVersion(app), {},
-                LedgerTestUtils::generateValidUniqueLedgerEntriesWithExclusions(
-                    {CONFIG_SETTING}, 8),
+                LedgerTestUtils::generateValidUniqueLedgerEntries(8),
```
We can't have `CONFIG_SETTINGS` types as a `deadEntry` or `initEntry` due to invariants, since `CONFIG_SETTING` can't be deleted and is guarenteed to already exist, so it can't be initialized either. However, it can (and does) get updated periodically, so I adjusted the test accordingly for slightly better coverage.

SirTyson commented on src/bucket/BucketManagerImpl.cpp:
```diff
@@ -599,21 +640,41 @@ BucketManagerImpl::getBucketIfExists(uint256 const& hash)
                    binToHex(hash), i->second->getFilename());
-        return i->second;
+
+        // Because BucketManger has an impl class, no public templated functions
+        // can be declared. This means we have to manually enforce types via
+        // `getLiveBucketByHash` and `getHotBucketByHash`, leading to this ugly
+        // cast.
+        auto ret = std::dynamic_pointer_cast<BucketT>(i->second);
```
Yeah unfortunately templated virtual functions are against the c++ spec, so the impl-class pattern is inconvenient here. I templated the inner BucketManagerImpl version, but we can only either have explicit return types with a messy dynamic cast, or just return a `BucketT` type in the `BucketManager` interface and have to cast on the caller side. I _think_ I might be able to get by with a `BucketT` level interface for futures and get rid of some of the constexpr branching, but we definitely need to be able to return `LiveBucket` explicitly in the   `getBucketByHash` case.

I'm fairly confident this is safe, since we have the Bucket type as part of meta so we're guaranteed to not have hash collision (except for empty buckets, but that should be fine). All that being said, I think separate `mSharedBuckets` would be useful so we could at least get rid of the dynamic cast, but we're stuck with the split `getLiveBucketByHash` interface I'm afraid.

SirTyson replied:
Hmm so I can't actually use a `BucketT` interface for futures either, but I can still have separate storage types in `BucketManager` to remove the cast. I think I can get rid of all the constexpr branching by changing the getBucket and getFutures interface to static getter that takes a `BucketManager&`. It's not a very elegant interface, but avoids removing the impl construct and removes most constexpr branching, so it's probably the least evil approach.

marta-lokhova replied:
note that this approach might be a bit annoying in tests, if we implement BucketList stub subclasses and such, but probably still worth exploring

SirTyson commented on src/bucket/BucketList.cpp:
```diff
@@ -511,16 +566,92 @@ BucketList::getSize() const
-                     std::vector<LedgerEntry> const& initEntries,
-                     std::vector<LedgerEntry> const& liveEntries,
-                     std::vector<LedgerKey> const& deadEntries)
+HotArchiveBucketList::addBatch(Application& app, uint32_t currLedger,
```
Good idea. A note for the reader, I used a parameter pack for the base class `addBatch`. While this isn't really necessary right now since both the `HotArchiveBucketList` and `LiveBucketList` have 3 input vector types, this is not true for the upcoming `ColdArchiveBucketList::addBatch` variant.

graydon replied:
I feel like this PR is growing into a bit of a C++ feature bingo card but I accept your parameter packs as appropriate nonetheless. obviously I asked for this.

SirTyson commented on src/catchup/IndexBucketsWork.cpp:
```diff
@@ -80,7 +80,8 @@ IndexBucketsWork::IndexWork::postWork()
             if (!self->mIndex)
             {
-                self->mIndex = BucketIndex::createIndex(
+                // TODO: Fix this when archive BucketLists assume state
```
The problem here is that we need to pipe the Bucket type into `IndexBucketsWork` so we can replace this `LiveBucket` with the proper templated `BucketT`. The problem is, `IndexBucketsWork` gets called on raw HAS reference files. The HAS doesn't currently support HotArchiveBucketList, so we can't pipe in the types yet. The umbrella issue is https://github.com/stellar/stellar-core/issues/4398.

Sorry if the outstanding TODOs aren't super clear, they're intended to be reminders for me in the not-so-obvious places I'll need to fix when I work on the `assumeState` path later.

SirTyson commented:
I should have addressed all the comments. I was able to refactor out almost all the `constexpr` branching, but had to use an ugly templated static interface for some `BucketManager` functions due to the `impl` construct. While it's not particularly pretty, this is probably the least worst option available given template constraints on virtual functions.

marta-lokhova commented on src/bucket/BucketManagerImpl.cpp:
```diff
@@ -486,11 +516,39 @@ BucketManagerImpl::renameBucketDirFile(std::filesystem::path const& src,
+                                 uint256 const& hash, MergeKey* mergeKey,
+                                 std::unique_ptr<BucketIndex const> index)
+{
+    auto& bmImpl = static_cast<BucketManagerImpl&>(bm);
```
i think we wouldn't need to do this if we didn't follow the impl-class pattern right?

marta-lokhova reviewed (commented):
Checkpointing here, as I'm still reviewing the change. Overall, really cool to see us getting closer to state archival! Thanks for putting it together. I left a few comments, mostly echoing @graydon's concerns about potential maintenance burden, and code readability. I think it's important we agree on an interface before we land this change.

SirTyson commented on src/bucket/BucketOutputIterator.cpp:
```diff
@@ -37,34 +42,93 @@ BucketOutputIterator::BucketOutputIterator(std::string const& tmpDir,
     if (protocolVersionStartsFrom(
             meta.ledgerVersion,
-            Bucket::FIRST_PROTOCOL_SUPPORTING_INITENTRY_AND_METAENTRY))
+            LiveBucket::FIRST_PROTOCOL_SUPPORTING_INITENTRY_AND_METAENTRY))
```
Yeah, but `HotArchiveBucket`  only occurs post p23 anyway. I don't think this check hurts us in any way and simplifies the code with one less `if constexpr`. I don't think it's worth special casing backwards compatibility code used only in reply for the new Bucket types.

SirTyson commented on src/main/Config.cpp:
```diff
@@ -1568,8 +1568,8 @@ Config::processConfig(std::shared_ptr<cpptoml::table> t)
-                     TESTING_STARTING_EVICTION_SCAN_LEVEL =
-                         readInt<uint32_t>(item, 1, BucketList::kNumLevels - 1);
+                     TESTING_STARTING_EVICTION_SCAN_LEVEL = readInt<uint32_t>(
+                         item, 1, LiveBucketList::kNumLevels - 1);
```
Yes

SirTyson replied:
So in future work this will be necessary, but for now they are identical, so I simplified the assignment a little.

SirTyson commented on src/bucket/BucketList.h:
```diff
@@ -398,11 +398,15 @@ class BucketListDepth
 
   public:
+    // Trivial pure virtual destructor to make this an abstract class
+    virtual ~BucketListBase() = 0;
```
No this was intended. I want the base BucketList class to be pure virtual, but I define all the functions it declares. so I define a trivial pure virtual destructor.

marta-lokhova replied:
oh, I didn't realize this was the _only_ pure virtual function in this class. This feels a bit hacky to me. An abstract class implies there's an interface subclasses should be implementing, or they can re-use the default implementation in the base class (which means base class functions would be virtual). Looking at the interface, why is `addBatch` not pure virtual for example? Forcing subclasses to implement destructors seems a bit weird, since it's a default implementation anyway.

marta-lokhova commented on src/bucket/BucketOutputIterator.cpp:
```diff
@@ -37,34 +42,93 @@ BucketOutputIterator::BucketOutputIterator(std::string const& tmpDir,
+            bme.metaEntry() = mMeta;
+            put(bme);
+        }
+        else
```
This still seems to be the case, as we don't explicitly require the right bucket type in the else case.

graydon commented on src/bucket/BucketIndexImpl.cpp:
```diff
@@ -126,35 +134,51 @@ BucketIndexImpl<IndexT>::BucketIndexImpl(BucketManager& bm,
-                //    load result is still correct even if the index has a few
-                //    deleted mappings.
-                if (be.type() == INITENTRY && key.type() == LIQUIDITY_POOL)
+                if constexpr (std::is_same_v<BucketEntryT, BucketEntry>)
```
```suggestion
                if constexpr (std::is_same_v<BucketEntryT, LiveBucket::EntryT>)
```

graydon commented on src/bucket/BucketOutputIterator.cpp:
```diff
@@ -37,34 +42,95 @@ BucketOutputIterator::BucketOutputIterator(std::string const& tmpDir,
+        }
+    }
+    else
     {
```
same else static-assert on bucket type here.
